src/linear_algebra/unitary_group.lean

```diff
@@ -0,0 +1,197 @@
+/-
+Copyright (c) 2021 Shing Tak Lam. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Author: Shing Tak Lam
+-/
+import linear_algebra.matrix
+import linear_algebra.nonsingular_inverse
+import data.complex.basic
+
+/-!
+# The Unitary Group
+
+This file defines elements of the unitary group `unitary_group n α`, where `α` is a `star_ring`.
+This consists of all `n` by `n` matrices with entries in `α` such that the star-transpose is its
+inverse. In addition, we define the group structure on `unitary_group n α`, and the embedding into
+the general linear group `general_linear_group α (n → α)`.
+
+We also define the orthogonal group `orthogonal_group n β`, where `β` is a `comm_ring`.
+
+## Main Definitions
+
+ * `matrix.unitary_group` is the type of matrices where the star-transpose is the inverse
+ * `matrix.unitary_group.group` is the group structure (under multiplication)
+ * `matrix.unitary_group.embedding_GL` is the embedding `unitary_group n α → GLₙ(α)`
+ * `matrix.orthogonal_group` is the type of matrices where the transpose is the inverse
+
+## References
+
+ * https://en.wikipedia.org/wiki/Unitary_group
+
+## Tags
+
+matrix group, group, unitary group, orthogonal group
+
+-/
+
+namespace matrix
+universes u v
+open linear_map
+open_locale matrix
+
+
+section
+
+variables (n : Type u) [fintype n] [decidable_eq n]
+variables (α : Type v) [comm_ring α] [star_ring α]
+
+/--
+`unitary_submonoid n α` is the submonoid consisting of all `n` by `n` matrices `A` where
+`star A ⬝ A = 1`.
+-/
+def unitary_submonoid : submonoid (matrix n n α) :=
+{ carrier := {A | star A ⬝ A = 1},
+  one_mem' := show star 1 ⬝ 1 = 1, by { rw [star_one, matrix.one_mul] },
+  mul_mem' := λ A B (hA : star A ⬝ A = 1) (hB : star B ⬝ B = 1), show star (A ⬝ B) ⬝ (A ⬝ B) = 1,
+    by { rwa [star_mul, ←matrix.mul_assoc, matrix.mul_assoc _ _ A, hA, matrix.mul_one] } }
```

Comment:
I wonder - can this be generalized to an arbitrary star-ring `A`, replacing `⬝` with `*`?

Reply:
I'm not so familiar with `star_ring`, so I'll let @shingtaklam1324 decide whether to do the generalization here or in a new PR.

Reply:
Yes, this is straightforward to generalize:
```lean
import algebra.star.basic
import group_theory.submonoid
import data.matrix.basic

#check 1

variables (R : Type*) [semiring R] [star_ring R]

def unitary_submonoid : submonoid R :=
{ carrier := {A | star A * A = 1},
  one_mem' := by simp,
  mul_mem' := λ A B (hA : star A * A = 1) (hB : star B * B = 1), show star (A * B) * (A * B) = 1,
    by rw [star_mul, mul_assoc, ←mul_assoc _ _ B, hA, one_mul, hB] }

variables (n : Type*) [fintype n] [decidable_eq n]
variables (α : Type*) [comm_ring α] [star_ring α]

example := unitary_submonoid (matrix n n α)
```

It's fine to keep lemmas specific to `unitary_submonoid (matrix n n α)` in this file, but I think we should use the general definition here.

Reply:
I'm happy to change `unitary_submonoid` to take any `star_ring R`. In fact, any `unitary_monoid` works. On the other hand, I would think that `unitary_group` should still be with matrices?

Reply:
Yes, that sounds good to me